### PR TITLE
STOR-455: support generic restore of volume using kdmp driver

### PR DESCRIPTION
**Pull request by ram-infrac**
**What type of PR is this?**
>feature

**What this PR does / why we need it**:
STOR-455: Support generic restore of snapshot using kdmp driver
- added storageclass mapping support for generic restore
 
**Does this PR change a user-facing CRD or CLI?**:
yes
```
apiVersion: stork.libopenstorage.org/v1alpha1
kind: ApplicationRestore
metadata:
  name: mysql-simple-restore
spec:
  backupName: mysql-test-bkp-1
  backupLocation: mysql-bl
  namespaceMapping:
    test: test
  replacePolicy: Delete
  storageClassMapping:
    mysql-sc: mysql-sc
```

**Is a release note needed?**:
yes
```release-note
Stork now support backup and restore of PX Proxy volume + CSI volume 
```

**Does this change need to be cherry-picked to a release branch?**:
yes, 2.8.0

Backup Spec: 
```
kind: ApplicationBackup
metadata:
  name: mysql-test-bkp-1
spec:
  backupLocation: mysql-bl
  backupType: generic
  # namespaces to be backed up
  namespaces:
  - test
  # What to do with the data in the objectstore when the backup object in k8s is deleted. Valid options are Delete and Retain
  reclaimPolicy: Delete
```

- scheduled mysql application in test namespace 
- added some records to db
- took generic backup by using above spec
- deleted mysql app and pvcs
- restore by giving sc mapping by using spec provided above
- verified pods are up and running
- verified data is present

### CODE REVIEW

File: drivers/volume/kdmp/kdmp.go [hunk 1]
```diff
@@ -231,26 +239,174 @@ func (k *kdmp) UpdateMigratedPersistentVolumeSpec(
...
+		dataExport, err := kdmpShedOps.Instance().GetDataExport(crName, vInfo.SourceNamespace)
+		if err != nil {
+			logrus.Errorf("failed to get DataExport CR: %v", err)
+			return volumeInfos, err
+		}
+		if dataExport.Status.TransferID == "" {
+			vInfo.Status = storkapi.ApplicationRestoreStatusInitial
+			vInfo.Reason = "Volume restore not started yet"
+		} else {
+			if isBackupActive(dataExport.Status) {
```
Comment (siva-portworx) [inline drivers/volume/kdmp/kdmp.go:391]: suggestion:It is good to change the function names isBackupActive and isBackupComplete as it is used in both backup and restore. Possible isDataExportActive and isDataExportCompleted?

File: drivers/volume/kdmp/kdmp.go [hunk 2]
```diff
@@ -231,26 +239,174 @@ func (k *kdmp) UpdateMigratedPersistentVolumeSpec(
...
+				// size for volume once support is available
+			} else if dataExport.Status.Status == kdmpapi.DataExportStatusFailed {
+				vInfo.Status = storkapi.ApplicationRestoreStatusFailed
+				vInfo.Reason = fmt.Sprintf("restore failed for volume: %v", dataExport.Status.Reason)
+			} else if isBackupComplete(dataExport.Status) {
+				vInfo.Status = storkapi.ApplicationRestoreStatusSuccessful
+				vInfo.Reason = "restore successful for volume"
+				// delete kdmp crs
+				if err := kdmpShedOps.Instance().DeleteDataExport(crName, restore.Namespace); err != nil {
+					logrus.Warnf("failed to delete data export CR: %v", err)
```
Comment (siva-portworx) [inline drivers/volume/kdmp/kdmp.go:404]: nit: It will be good to have DataExport CR name as well in the debug.

File: drivers/volume/kdmp/kdmp.go [hunk 3]
```diff
@@ -231,26 +239,174 @@ func (k *kdmp) UpdateMigratedPersistentVolumeSpec(
...
+			vInfo.Reason = "Volume restore not started yet"
+		} else {
+			if isBackupActive(dataExport.Status) {
+				vInfo.Status = storkapi.ApplicationRestoreStatusInProgress
+				vInfo.Reason = "Volume restore is in progress. BytesDone"
+				// TODO: KDMP does not return size right now, we will need to fill up
+				// size for volume once support is available
+			} else if dataExport.Status.Status == kdmpapi.DataExportStatusFailed {
+				vInfo.Status = storkapi.ApplicationRestoreStatusFailed
+				vInfo.Reason = fmt.Sprintf("restore failed for volume: %v", dataExport.Status.Reason)
```
Comment (siva-portworx) [inline drivers/volume/kdmp/kdmp.go:398]: nit: Same here, Volume name can be part of debug statement.

File: drivers/volume/kdmp/kdmp.go [hunk 4]
```diff
@@ -231,26 +239,174 @@ func (k *kdmp) UpdateMigratedPersistentVolumeSpec(
...
+			Namespace:  volBackup.Namespace,
+			APIVersion: KdmpAPIVersion,
+		}
+		dataExport.Spec.Destination = kdmpapi.DataExportObjectReference{
+			Kind:       PVCKind,
+			Name:       bkpvInfo.PersistentVolumeClaim,
+			Namespace:  bkpvInfo.Namespace,
+			APIVersion: "v1",
+		}
+		if _, err := kdmpShedOps.Instance().CreateDataExport(dataExport); err != nil {
```
Comment (siva-portworx) [inline drivers/volume/kdmp/kdmp.go:355]: Should we check for already exist error and ignore the error, when any retry happens for a volume.

File: drivers/volume/kdmp/kdmp.go [hunk 5]
```diff
@@ -231,26 +239,174 @@ func (k *kdmp) UpdateMigratedPersistentVolumeSpec(
...
+		volBackup.Namespace = bkpvInfo.Namespace
+		volBackup.Spec.BackupLocation = kdmpapi.DataExportObjectReference{
+			Kind:       reflect.TypeOf(storkapi.BackupLocation{}).Name(),
+			Name:       restore.Spec.BackupLocation,
+			Namespace:  restore.Namespace,
+			APIVersion: StorkAPIVersion,
+		}
+		volBackup.Spec.Repository = fmt.Sprintf("%s/%s-%s/", prefixRepo, bkpvInfo.Namespace, bkpvInfo.PersistentVolumeClaim)
+		volBackup.Status.SnapshotID = bkpvInfo.BackupID
+		if _, err := kdmpShedOps.Instance().CreateVolumeBackup(volBackup); err != nil {
```
Comment (siva-portworx) [inline drivers/volume/kdmp/kdmp.go:330]: Should we check for already exist error and ignore the error, when any retry happens for a volume.

**Review by siva-portworx — APPROVED** [review]

File: drivers/volume/kdmp/kdmp.go (hunk 1 above)
Comment (ram-infrac) [reply drivers/volume/kdmp/kdmp.go]: yes, updated

File: drivers/volume/kdmp/kdmp.go (hunk 4 above)
Comment (ram-infrac) [reply drivers/volume/kdmp/kdmp.go:355]: applicationcontroller will only retry restore if driver send `Busy` error backup to controller

**Pull request merged**